File: port_scanner.py

```python
#!/usr/bin/env python3
# port_scanner.py
import argparse
import socket
import threading
import time
import json
import csv
import sys
from datetime import datetime
from colorama import init, Fore, Style

init(autoreset=True)

class TCPScanner:
    def __init__(self, host, ports, timeout=2, threads=10, verbose=False, color=True):
        self.host = host
        try:
            self.ip = socket.gethostbyname(host)
        except:
            self.ip = host
        self.ports = self.parse_ports(ports)
        self.timeout = timeout
        self.threads = min(threads, len(self.ports))
        self.verbose = verbose
        self.color = color and sys.stdout.isatty()
        self.results = []
        self.lock = threading.Lock()
        self._stop = False

    def parse_ports(self, port_str):
        ports = set()
        for part in port_str.split(','):
            if '-' in part:
                start, end = map(int, part.split('-'))
                ports.update(range(start, end+1))
            else:
                ports.add(int(part))
        return sorted(ports)

    def scan_port(self, port):
        try:
            sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
            sock.settimeout(self.timeout)
            result = sock.connect_ex((self.ip, port))
            sock.close()
            if result == 0:
                return {"port": port, "state": "open", "banner": ""}
            else:
                return {"port": port, "state": "closed", "banner": ""}
        except socket.timeout:
            return {"port": port, "state": "filtered", "banner": ""}
        except Exception as e:
            return {"port": port, "state": "error", "banner": str(e)}

    def worker(self, port_list):
        for port in port_list:
            if self._stop:
                break
            res = self.scan_port(port)
            with self.lock:
                self.results.append(res)
                if self.verbose:
                    self.print_result(res)

    def scan(self):
        print(f"Scanning {self.host} ({self.ip})...")
        start_time = time.time()
        # Разбиваем порты на чанки для потоков
        chunk_size = max(1, len(self.ports) // self.threads)
        chunks = [self.ports[i:i+chunk_size] for i in range(0, len(self.ports), chunk_size)]
        threads = []
        for chunk in chunks:
            t = threading.Thread(target=self.worker, args=(chunk,))
            t.start()
            threads.append(t)
        # Ожидаем завершения всех потоков
        for t in threads:
            t.join()
        elapsed = time.time() - start_time
        open_count = len([r for r in self.results if r['state'] == 'open'])
        print(f"\nScan completed in {elapsed:.2f}s. Found {open_count} open ports.")
        return self.results

    def print_result(self, res):
        if self.color:
            if res['state'] == 'open':
                color = Fore.GREEN
            elif res['state'] == 'closed':
                color = Fore.RED
            elif res['state'] == 'filtered':
                color = Fore.YELLOW
            else:
                color = Fore.WHITE
            print(f"{color}Port {res['port']}/tcp  {res['state']}{' '*10} {res['banner'][:30]}")
        else:
            print(f"Port {res['port']}/tcp  {res['state']}  {res['banner'][:30]}")

    def export_json(self, filename):
        with open(filename, 'w') as f:
            json.dump({"host": self.host, "ip": self.ip, "results": self.results}, f, indent=2)

    def export_csv(self, filename):
        with open(filename, 'w', newline='') as f:
            writer = csv.writer(f)
            writer.writerow(["port", "state", "banner"])
            for r in self.results:
                writer.writerow([r['port'], r['state'], r['banner']])

def main():
    parser = argparse.ArgumentParser(description="TCP Port Scanner")
    parser.add_argument("--host", required=True, help="Target host (IP or domain)")
    parser.add_argument("--ports", default="22,80,443,1-1024", help="Ports to scan (e.g., 80,443,1-1000)")
    parser.add_argument("--timeout", type=float, default=2.0, help="Connection timeout in seconds")
    parser.add_argument("--threads", type=int, default=10, help="Number of threads")
    parser.add_argument("--json", help="Export results to JSON file")
    parser.add_argument("--csv", help="Export results to CSV file")
    parser.add_argument("--verbose", action="store_true", help="Verbose output")
    parser.add_argument("--no-color", action="store_true", help="Disable colored output")
    args = parser.parse_args()

    scanner = TCPScanner(
        host=args.host,
        ports=args.ports,
        timeout=args.timeout,
        threads=args.threads,
        verbose=args.verbose,
        color=not args.no_color
    )
    results = scanner.scan()
    if args.json:
        scanner.export_json(args.json)
        print(f"Results exported to {args.json}")
    if args.csv:
        scanner.export_csv(args.csv)
        print(f"Results exported to {args.csv}")

if __name__ == "__main__":
    main()

File: port_scanner.js
#!/usr/bin/env node
// port_scanner.js
const { program } = require('commander');
const net = require('net');
const fs = require('fs');
const chalk = require('chalk');

class TCPScanner {
    constructor(host, ports, timeout, threads, verbose, color) {
        this.host = host;
        this.ip = this.resolveIP(host);
        this.ports = this.parsePorts(ports);
        this.timeout = timeout * 1000;
        this.threads = Math.min(threads, this.ports.length);
        this.verbose = verbose;
        this.color = color && process.stdout.isTTY;
        this.results = [];
        this.lock = false;
        this._stop = false;
    }

    resolveIP(host) {
        try {
            return require('dns').lookupSync(host).address;
        } catch (e) {
            return host;
        }
    }

    parsePorts(portStr) {
        const ports = new Set();
        portStr.split(',').forEach(part => {
            if (part.includes('-')) {
                const [start, end] = part.split('-').map(Number);
                for (let p = start; p <= end; p++) ports.add(p);
            } else {
                ports.add(Number(part));
            }
        });
        return Array.from(ports).sort((a,b) => a-b);
    }

    scanPort(port) {
        return new Promise((resolve) => {
            const socket = new net.Socket();
            let resolved = false;
            const timer = setTimeout(() => {
                if (!resolved) {
                    resolved = true;
                    socket.destroy();
                    resolve({ port, state: 'filtered', banner: '' });
                }
            }, this.timeout);

            socket.on('connect', () => {
                if (!resolved) {
                    resolved = true;
                    clearTimeout(timer);
                    socket.destroy();
                    resolve({ port, state: 'open', banner: '' });
                }
            });

            socket.on('error', (err) => {
                if (!resolved) {
                    resolved = true;
                    clearTimeout(timer);
                    socket.destroy();
                    let state = 'error';
                    if (err.code === 'ECONNREFUSED') state = 'closed';
                    else if (err.code === 'ETIMEDOUT') state = 'filtered';
                    resolve({ port, state, banner: err.message });
                }
            });

            socket.connect(port, this.ip);
        });
    }

    async scanChunk(chunk) {
        const promises = chunk.map(p => this.scanPort(p));
        const results = await Promise.all(promises);
        for (const res of results) {
            this.results.push(res);
            if (this.verbose) this.printResult(res);
        }
    }

    async scan() {
        console.log(`Scanning ${this.host} (${this.ip})...`);
        const start = Date.now();
        const chunkSize = Math.ceil(this.ports.length / this.threads);
        const chunks = [];
        for (let i = 0; i < this.ports.length; i += chunkSize) {
            chunks.push(this.ports.slice(i, i+chunkSize));
        }
        const promises = chunks.map(chunk => this.scanChunk(chunk));
        await Promise.all(promises);
        const elapsed = (Date.now() - start) / 1000;
        const open = this.results.filter(r => r.state === 'open').length;
        console.log(`\nScan completed in ${elapsed.toFixed(2)}s. Found ${open} open ports.`);
        return this.results;
    }

    printResult(res) {
        if (this.color) {
            let colorFn = chalk.white;
            if (res.state === 'open') colorFn = chalk.green;
            else if (res.state === 'closed') colorFn = chalk.red;
            else if (res.state === 'filtered') colorFn = chalk.yellow;
            console.log(colorFn(`Port ${res.port}/tcp  ${res.state}  ${res.banner.slice(0,30)}`));
        } else {
            console.log(`Port ${res.port}/tcp  ${res.state}  ${res.banner.slice(0,30)}`);
        }
    }

    exportJSON(filename) {
        fs.writeFileSync(filename, JSON.stringify({ host: this.host, ip: this.ip, results: this.results }, null, 2));
    }

    exportCSV(filename) {
        const lines = ['port,state,banner'];
        this.results.forEach(r => {
            lines.push(`${r.port},${r.state},"${r.banner}"`);
        });
        fs.writeFileSync(filename, lines.join('\n'));
    }
}

program
    .requiredOption('--host <host>', 'Target host')
    .option('--ports <ports>', 'Ports to scan', '22,80,443,1-1024')
    .option('--timeout <seconds>', 'Timeout in seconds', parseFloat, 2)
    .option('--threads <number>', 'Number of threads', parseInt, 10)
    .option('--json <file>', 'Export to JSON')
    .option('--csv <file>', 'Export to CSV')
    .option('--verbose', 'Verbose output')
    .option('--no-color', 'Disable color')
    .parse(process.argv);

const opts = program.opts();
const scanner = new TCPScanner(
    opts.host,
    opts.ports,
    opts.timeout,
    opts.threads,
    opts.verbose,
    !opts.noColor
);
scanner.scan().then(() => {
    if (opts.json) {
        scanner.exportJSON(opts.json);
        console.log(`Results exported to ${opts.json}`);
    }
    if (opts.csv) {
        scanner.exportCSV(opts.csv);
        console.log(`Results exported to ${opts.csv}`);
    }
});
